flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/read/HiveBulkFormatAdapter.java

```diff
@@ -0,0 +1,291 @@
+
+	private static final long serialVersionUID = 1L;
+
+	private static final Logger LOG = LoggerFactory.getLogger(HiveBulkFormatAdapter.class);
+
+	// schema evolution configs are not available in older versions of IOConstants, let's define them ourselves
+	private static final String SCHEMA_EVOLUTION_COLUMNS = "schema.evolution.columns";
+	private static final String SCHEMA_EVOLUTION_COLUMNS_TYPES = "schema.evolution.columns.types";
+
+	private final JobConfWrapper jobConfWrapper;
+	private final List<String> partitionKeys;
+	private final String[] fieldNames;
+	private final DataType[] fieldTypes;
+	// indices of fields to be returned, with projection applied (if any)
+	private final int[] selectedFields;
+	private final String hiveVersion;
+	private final HiveShim hiveShim;
+	private final DataType producedDataType;
+	private final boolean useMapRedReader;
+	// We should limit the input read count of the splits, -1 represents no limit.
+	private final long limit;
+
+	public HiveBulkFormatAdapter(JobConfWrapper jobConfWrapper, List<String> partitionKeys, String[] fieldNames, DataType[] fieldTypes,
+			int[] selectedFields, String hiveVersion, DataType producedDataType, boolean useMapRedReader, long limit) {
+		this.jobConfWrapper = jobConfWrapper;
+		this.partitionKeys = partitionKeys;
+		this.fieldNames = fieldNames;
+		this.fieldTypes = fieldTypes;
+		this.selectedFields = selectedFields;
+		this.hiveVersion = hiveVersion;
+		this.hiveShim = HiveShimLoader.loadHiveShim(hiveVersion);
+		this.producedDataType = producedDataType;
+		this.useMapRedReader = useMapRedReader;
+		this.limit = limit;
+	}
+
+	@Override
+	public Reader<RowData> createReader(Configuration config, HiveSourceSplit split)
+			throws IOException {
+		return createBulkFormatForSplit(split).createReader(config, split);
+	}
+
+	@Override
+	public Reader<RowData> restoreReader(Configuration config, HiveSourceSplit split) throws IOException {
+		return createBulkFormatForSplit(split).restoreReader(config, split);
+	}
+
+	@Override
+	public boolean isSplittable() {
+		return true;
+	}
+
+	@Override
+	public TypeInformation<RowData> getProducedType() {
+		return (TypeInformation<RowData>) TypeInfoDataTypeConverter.fromDataTypeToTypeInfo(producedDataType);
+	}
+
+	private BulkFormat<RowData, ? super HiveSourceSplit> createBulkFormatForSplit(HiveSourceSplit split) {
+		if (!useMapRedReader && useParquetVectorizedRead(split.getHiveTablePartition())) {
+			// TODO: need a way to support limit push down
+			return ParquetColumnarRowInputFormat.createPartitionedFormat(
+					jobConfWrapper.conf(),
+					(RowType) producedDataType.getLogicalType(),
+					partitionKeys,
+					jobConfWrapper.conf().get(HiveConf.ConfVars.DEFAULTPARTITIONNAME.varname,
+							HiveConf.ConfVars.DEFAULTPARTITIONNAME.defaultStrVal),
+					(PartitionValueConverter) (colName, valStr, type) -> split.getHiveTablePartition().getPartitionSpec().get(colName),
```

Comment:
This will fail when partition values can not be found in path.
I create https://github.com/apache/flink/pull/13919 to refactor this.